### PR TITLE
Looping audio

### LINKED ISSUE #129: [MLH Fellow] [Audio] Add `loop` augmentation
## 🚀 Feature
Implement a new augmentation called `loop` which takes in an audio file/numpy array and returns it looped a given `n` number of times.

## To do
- Read our [contributing](https://github.com/facebookresearch/AugLy/blob/main/CONTRIBUTING.md#adding-a-new-augmentation) guidelines for all the steps to add a new augmentation
- Note: for this augmentation you will also need to define how the matching segments will change [here](https://github.com/facebookresearch/AugLy/blob/main/augly/audio/utils.py#L136). This should work similarly to how the segments are computed for the video `loop` augmentation [here](https://github.com/facebookresearch/AugLy/blob/main/augly/video/helpers/metadata.py#L176).
- Make sure the audio tests (including the new one you have added) pass: `python -m unittest discover -s augly/tests/audio_tests/ -p "*"`

### PR DESCRIPTION
**Pull request by prasad-aditya**
## Related Issue
Fixes #129 

## Summary
- [x] I have read CONTRIBUTING.md to understand how to contribute to this repository :)

<Please summarize what you are trying to achieve, what changes you made, and how they achieve the desired result.>
Preliminary loop augmentation for the audio modality, yet to add stereo `test_loop_audio.wave`

## Unit Tests
If your changes touch the `audio` module, please run all of the `audio` tests and paste the output here. Likewise for `image`, `text`, & `video`. If your changes could affect behavior in multiple modules, please run the tests for all potentially affected modules. If you are unsure of which modules might be affected by your changes, please just run all the unit tests.

### Audio
```bash
python -m unittest discover -s augly/tests/audio_tests/ -p "*"

......./home/adityaprasad/.local/lib/python3.8/site-packages/librosa/util/utils.py:2099: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.dtype(np.float): np.complex,
/home/adityaprasad/.local/lib/python3.8/site-packages/librosa/util/utils.py:2099: DeprecationWarning: `np.complex` is a deprecated alias for the builtin `complex`. To silence this warning, use `complex` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.complex128` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.dtype(np.float): np.complex,
/home/adityaprasad/.local/lib/python3.8/site-packages/librosa/util/utils.py:869: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  mag = np.abs(S).astype(np.float)
........../home/adityaprasad/.local/lib/python3.8/site-packages/librosa/core/spectrum.py:1223: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  time_steps = np.arange(0, D.shape[1], rate, dtype=np.float)
/home/adityaprasad/.local/lib/python3.8/site-packages/librosa/util/utils.py:2099: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.dtype(np.float): np.complex,
/home/adityaprasad/.local/lib/python3.8/site-packages/librosa/util/utils.py:2099: DeprecationWarning: `np.complex` is a deprecated alias for the builtin `complex`. To silence this warning, use `complex` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.complex128` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  np.dtype(np.float): np.complex,
/home/adityaprasad/.local/lib/python3.8/site-packages/librosa/core/spectrum.py:1223: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  time_steps = np.arange(0, D.shape[1], rate, dtype=np.float)
/home/adityaprasad/.local/lib/python3.8/site-packages/librosa/util/utils.py:869: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  mag = np.abs(S).astype(np.float)
...................................................
----------------------------------------------------------------------
Ran 68 tests in 4.976s

OK
```

### Image
```bash
python -m unittest discover -s augly/tests/image_tests/ -p "*_test.py"
# Or `python -m unittest discover -s augly/tests/image_tests/ -p "*.py"` to run pytorch test too (must install `torchvision` to run)
```

### Text
```bash
python -m unittest discover -s augly/tests/text_tests/ -p "*"
```

### Video
```bash
python -m unittest discover -s augly/tests/video_tests/ -p "*"
```

### All
```bash
python -m unittest discover -s augly/tests/ -p "*"
```

## Other testing

If applicable, test your changes and paste the output here. For example, if your changes affect the requirements/installation, then test installing augly in a fresh conda env, then make sure you are able to import augly & run the unit test

### CODE REVIEW

**Review by iAdityaEmpire — COMMENTED** [review]
Have yet to add an `sf.write()` for the stereo audio, slightly confused on how the implementation would be different from the mono version (which was the default)

File: augly/audio/functional.py [hunk 1]
```diff
@@ -67,7 +67,8 @@ def add_background_noise(
     if background_audio is None:
         background_audio = random_generator.standard_normal(audio.shape)
     else:
-        background_audio, _ = audutils.validate_and_load_audio(background_audio, 1)
+        background_audio, _ = audutils.validate_and_load_audio(
```
Comment (zpapakipos) [inline augly/audio/functional.py:70]: Please get rid of these formatting changes which are unrelated to your PR :) It makes it harder to see your actual changes; we know we need to reformat all files using `black`, we will do it soon in a separate PR.

File: augly/audio/intensity.py [hunk 2]
```diff
@@ -201,3 +201,10 @@ def time_stretch_intensity(rate: float = 1.5, **kwargs) -> float:
 
 def to_mono_intensity(metadata: Dict[str, Any], **kwargs) -> float:
     return 0.0 if metadata["src_num_channels"] == 1 else 100.0
+
+# unsure as to how intensity works
+def loop_audio_intensity(n: int = 1, **kwargs) -> float:
```
Comment (zpapakipos) [inline augly/audio/intensity.py:206]: Why not implement this like [the one](https://github.com/facebookresearch/AugLy/blob/main/augly/video/helpers/intensity.py#L160) for `loop` in video?

**Review by zpapakipos — CHANGES_REQUESTED** [review]
> Have yet to add an sf.write() for the stereo audio, slightly confused on how the implementation would be different from the mono version (which was the default)

You can just call [`ret_and_save_audio`](https://github.com/facebookresearch/AugLy/blob/main/augly/audio/utils.py#L49) from `augly.audio.utils`! It works for any number of channels & different audio formats.

**Comment by iAdityaEmpire** [comment]
@zpapakipos @jbitton The `test_python` check is apparently failing on the installation of `pytorch`, but my local tests are passing. Is this something I should somehow fix?

File: augly/audio/functional.py [hunk 3]
```diff
@@ -1269,3 +1269,44 @@ def to_mono(
     )
 
     return audutils.ret_and_save_audio(aug_audio, output_path, sample_rate)
+
+
+def loop_audio(
```
Comment (jbitton) [inline augly/audio/functional.py:1274]: two tiny nits:
(a) let's name this function `loop` for consistency with other libraries :)
(b) instead of adding this function to the end of `functional.py`, could you insert it in alphabetically with respect to the other function names in this file?

File: augly/audio/functional.py [hunk 4]
```diff
@@ -1269,3 +1269,44 @@ def to_mono(
...
+    Takes in an audio file/numpy array and returns it looped a given n number of times
+    @param audio: the path to the audio or a variable of type np.ndarray that
+        will be augmented
+    @param n: the number of times the video will be looped
+    @param output_path: the path in which the resulting audio will be stored. If None,
+        the resulting np.ndarray will still be returned
+    @param metadata: if set to be a list, metadata about the function execution
+        including its name, the source & dest duration, sample rates, etc. will be
+        appended to the inputted list. If set to None, no metadata will be appended
+    @returns: the augmented audio array
```
Comment (jbitton) [inline augly/audio/functional.py:1291]:
```suggestion
    Loops the audio 'n' times

    @param audio: the path to the audio or a variable of type np.ndarray that
        will be augmented
    
    @param sample_rate: the audio sample rate of the inputted audio

    @param n: the number of times the video will be looped

    @param output_path: the path in which the resulting audio will be stored. If None,
        the resulting np.ndarray will still be returned

    @param metadata: if set to be a list, metadata about the function execution
        including its name, the source & dest duration, sample rates, etc. will be
        appended to the inputted list. If set to None, no metadata will be appended

    @returns: the augmented audio array and sample rate
```
making the docstring more consistent with others in the library (formatting & language-wise) and adding in missing params

File: augly/audio/functional.py [hunk 5]
```diff
@@ -1269,3 +1269,44 @@ def to_mono(
...
+    """
+    assert isinstance(n, int) and n >= 0, "Expected 'n' to be a nonnegative integer"
+    audio, sample_rate = audutils.validate_and_load_audio(audio, sample_rate)
+    aug_audio = audio
+    for i in range(n):
+        aug_audio = np.append(aug_audio, audio, axis=(0 if audio.ndim == 1 else 1))
+    # or possibly:
+    # aug_audio = np.repeat(audio, 7, axis=0)
+    # aug_audio = aug_audio.flatten()
```
Comment (jbitton) [inline augly/audio/functional.py:1300]:
```suggestion
    audio, sample_rate = audutils.validate_and_load_audio(audio, sample_rate)

    aug_audio = audio
    for i in range(n):
        aug_audio = np.append(aug_audio, audio, axis=(0 if audio.ndim == 1 else 1))
```
remove commented out code and add newline to separate parts of the function :)

File: augly/audio/functional.py [hunk 6]
```diff
@@ -1269,3 +1269,44 @@ def to_mono(
...
+
+    audutils.get_metadata(
+        metadata=metadata,
+        function_name="loop_audio",
```
Comment (jbitton) [inline augly/audio/functional.py:1304]: +1 on changing function name :)

File: augly/audio/functional.py [hunk 7]
```diff
@@ -1269,3 +1269,44 @@ def to_mono(
...
+        audio=audio,
+        sample_rate=sample_rate,
+        dst_audio=aug_audio,
+        dst_sample_rate=sample_rate,
+        output_path=output_path,
```
Comment (jbitton) [inline augly/audio/functional.py:1309]: let's add `n` as a kwarg so we can log what value of `n` we used in the function!

File: augly/audio/intensity.py [hunk 8]
```diff
@@ -201,3 +201,9 @@ def time_stretch_intensity(rate: float = 1.5, **kwargs) -> float:
 
 def to_mono_intensity(metadata: Dict[str, Any], **kwargs) -> float:
     return 0.0 if metadata["src_num_channels"] == 1 else 100.0
+
+
+def loop_audio_intensity(n: int = 1, **kwargs) -> float:
+    assert isinstance(n, int) and n >= 1, "n must be a positive integer"
```
Comment (jbitton) [inline augly/audio/intensity.py:207]: you'll need to change this assert to match the one in `functional.py` (also it would be great if you can change this func name/reorder this function to be in alphabetical order, similar to the suggestion i made in `functional.py`)

File: augly/audio/transforms.py [hunk 9]
```diff
@@ -884,3 +884,35 @@ def apply_transform(
         @returns: the augmented audio array and sample rate
         """
         return F.to_mono(audio, sample_rate, metadata=metadata)
+
+
+class LoopAudio(BaseTransform):
```
Comment (jbitton) [inline augly/audio/transforms.py:889]: +1 to class name change
+1 to reordering such that it's in alphabetical order

File: augly/audio/transforms.py [hunk 10]
```diff
@@ -884,3 +884,35 @@ def apply_transform(
...
+        self.n = n
+
+    def apply_transform(
+        self,
+        audio: np.ndarray,
+        sample_rate: int,
+        metadata: Optional[List[Dict[str, Any]]] = None,
+    ) -> Tuple[np.ndarray, int]:
+        """
+        Converts the audio from stereo to mono by averaging samples across channels
```
Comment (jbitton) [inline augly/audio/transforms.py:906]: change description to match the new description suggested in functional.py :)

File: augly/tests/audio_tests/functional_unit_test.py [hunk 11]
```diff
@@ -69,6 +69,9 @@ def test_time_stretch(self):
     def test_to_mono(self):
         self.evaluate_function(audaugs.to_mono)
 
+    def test_loop_audio(self):
```
Comment (jbitton) [inline augly/tests/audio_tests/functional_unit_test.py:72]: +1 to name change
+1 to reordering such that it's in alphabetical order

File: augly/tests/audio_tests/intensity_unit_test.py [hunk 12]
```diff
@@ -93,3 +93,7 @@ def test_time_stretch_intensity(self):
     def test_to_mono_intensity(self):
         intensity = audaugs.to_mono_intensity(metadata={"src_num_channels": 1})
         self.assertAlmostEqual(intensity, 0.0)
+
+    def test_loop_audio_intensity(self):
```
Comment (jbitton) [inline augly/tests/audio_tests/intensity_unit_test.py:97]: +1 to name change
+1 to reordering such that it's in alphabetical order

File: augly/tests/audio_tests/transforms_unit_test.py [hunk 13]
```diff
@@ -133,6 +133,9 @@ def test_TimeStretch(self):
     def test_ToMono(self):
         self.evaluate_class(audaugs.ToMono(), fname="to_mono")
 
+    def test_LoopAudio(self):
```
Comment (jbitton) [inline augly/tests/audio_tests/transforms_unit_test.py:136]: +1 to name change
+1 to reordering such that it's in alphabetical order

File: augly/utils/expected_output/audio_tests/expected_metadata.json [hunk 14]
```diff
@@ -649,5 +649,29 @@
           }
         ]
       }
+    ],
+    "loop_audio": [
+      { 
+        "name": "loop_audio",
```
Comment (jbitton) [inline augly/utils/expected_output/audio_tests/expected_metadata.json:655]: you'll need to change these too :)

**Review by jbitton — COMMENTED** [review]
Overall this looks good to me! So sorry that you had to go through all those merge conflicts / essentially re-do the PR. Good news is that you're some nits away from getting your first AugLy PR landed!!!

That issue in the GitHub workflow shouldn't be your fault. Once you address these nits etc, let's see if the tests now pass / if it was a transient error. If not, @zpapakipos and I will look into it more -- it could be that one of our dependencies got a new update and now something isn't working.

File: augly/audio/functional.py (hunk 5 above)
Comment (zpapakipos) [reply augly/audio/functional.py:1300]: Agree with this change, but one more small nit: change `for i in range(n)` to `for _ in range(n)` since the var `i` is never used in the loop :)

File: augly/audio/__init__.py [hunk 15]
```diff
@@ -77,7 +80,7 @@
     "high_pass_filter",
     "insert_in_background",
     "invert_channels",
-    "low_pass_filter",
+    "loop" "low_pass_filter",
```
Comment (jbitton) [inline augly/audio/__init__.py:83]:
```suggestion
    "loop",
    "low_pass_filter",
```
forgot comma and newline

File: augly/audio/functional.py [hunk 16]
```diff
@@ -617,6 +617,52 @@ def invert_channels(
...
+def loop(
+    audio: Union[str, np.ndarray],
+    sample_rate: int = DEFAULT_SAMPLE_RATE,
+    n: int = 1,
+    output_path: Optional[str] = None,
+    metadata: Optional[List[Dict[str, Any]]] = None,
+) -> Tuple[np.ndarray, int]:
+    """
+    Loops the audio 'n' times
+    @param audio: the path to the audio or a variable of type np.ndarray that
```
Comment (jbitton) [inline augly/audio/functional.py:629]:
```suggestion
    Loops the audio 'n' times

    @param audio: the path to the audio or a variable of type np.ndarray that
```

File: augly/audio/intensity.py [hunk 17]
```diff
@@ -100,6 +100,12 @@ def invert_channels_intensity(metadata: Dict[str, Any], **kwargs) -> float:
     return 0.0 if metadata["src_num_channels"] == 1 else 100.0
 
 
+def loop_intensity(n: int = 1, **kwargs) -> float:
+    assert isinstance(n, int) and n >= 0, "Expected 'n' to be a nonnegative integer"
+    max_num_loops = 100
```
Comment (jbitton) [inline augly/audio/intensity.py:105]:
```suggestion
    assert isinstance(n, int) and n >= 0, "Expected 'n' to be a nonnegative integer"

    max_num_loops = 100
```

File: augly/audio/transforms.py [hunk 18]
```diff
@@ -450,6 +450,38 @@ def apply_transform(
...
+
+        @param audio: the path to the audio or a variable of type np.ndarray that
+            will be augmented
+
+        @param sample_rate: the audio sample rate of the inputted audio
+
+        @param metadata: if set to be a list, metadata about the function execution
+            including its name, the source & dest duration, sample rates, etc. will be
+            appended to the inputted list. If set to None, no metadata will be appended
+            @returns: the augmented audio array and sample rate
```
Comment (jbitton) [inline augly/audio/transforms.py:480]:
```suggestion
        @param metadata: if set to be a list, metadata about the function execution
            including its name, the source & dest duration, sample rates, etc. will be
            appended to the inputted list. If set to None, no metadata will be appended

        @returns: the augmented audio array and sample rate
```

param was accidentally indented & added newline

**Review by jbitton — APPROVED** [review]
Overall LGTM! A couple teeny tiny nits (mostly just adding newlines) and we're done! Btw, I ran your code internally and it passes :)

**Comment by iAdityaEmpire** [comment]
@jbitton Got it! The new commits should be there now.
